### PR TITLE
FEATURE: Allow RoutePart handlers to point to external URIs with query

### PR DESCRIPTION
**Pull request by bwaidelich**
With this change, RoutePart handlers can define all relevant URI features
including query string and fragment in order to point to external URIs:

```php
class SomeRoutePartHandler extends DynamicRoutePart {

    protected function resolveValue($value) {
        return new ResolveResult('', UriConstraints::fromUri(new Uri('https://neos.io:8080/some/path?some[query]=string#some-fragment')));
    }
}
```

Background:

This is a preparation to fully support cross-domain routing. Also, for Neos, this will be
required in order to deal with shortcut nodes pointing to external URLs within the routing context

Related: #1126, neos/neos-development-collection#3020
Fixes: #2140

### CODE REVIEW

File: Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php [hunk 1]
```diff
@@ -229,6 +289,7 @@ public function withPathSuffix(string $pathSuffix, bool $prepend = false): self
      * Returns the URI path constraint, which consists of the path and query string parts, or NULL if none was set
      *
      * @return string|null
+     * @deprecated With Flow 7.0, use toUri()->getPath() instead. @see toUri()
```
Comment (albe) [inline Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php:292]: Does this really return the same value as `toUri()->getPath()`? If so, can we just replace the method body with that?

**Review by albe — COMMENTED** [review]
Looks good by reading, but left two comments

File: Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php [hunk 2]
```diff
@@ -14,7 +14,9 @@
 use Neos\Flow\Annotations as Flow;
 use GuzzleHttp\Psr7\Uri;
 use Neos\Flow\Http\Helper\UriHelper;
+use Neos\Utility\Arrays;
 use Psr\Http\Message\UriInterface;
+use function GuzzleHttp\Psr7\parse_query;
```
[the remark this answers is not in the document]
Comment (bwaidelich) [reply Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php:19]: `parse_url` parses an URL and returns an assoc array of its part.
`parse_query` parses a query string into an assoc array.

With the former `?foo[bar]=baz` would return
```
{"query": "foo[bar]=baz"}
```
with the latter
```
{"foo": {"bar": "baz"}
```

> is the reason we don't also use guzzle build_query over php http_build_query?

These two serve the same purpose but the one from guzzle removes duplicates IIRC. In this part we already merged params so there won't be any duplicates so I would keep the c function for performance reasons

File: Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php:292]: Good call, I'll give that a try

Comment (bwaidelich) [reply Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php]: OK, I checked this again, but it's not exactly the same.
`toUri()->getPath()` will return a string.
This obsolete method returns a nullable string and it also adds the query string (as mentioned in the doc comment).
I adjusted the `@deprecated` annotation accordingly.

In general I don't think that this method is actually used outside of the core and there is no big harm in keeping it for a while. It's just not very beautiful :)

File: Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (kdambekalns) [reply Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php:19]: To parse the query into an array, why not use https://www.php.net/manual/en/function.parse-str.php?

[the remark this answers is not in the document]
Comment (albe) [reply Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php:19]: Ah right, there is the oddly named counterpart

**Review by albe — APPROVED** [review]
Looks good. The guzzle `parse_query` could be replaced as in the comment, but it's not pulling a new dependency in so I'm fine with that in the end.

**Comment by bwaidelich** [comment]
Thanks for the reviews
I remember that there was a good reason to use Guzzle `parse_query` over PHPs built-in `parse_str` but I'll look it up again and comment the code accordingly

**Comment by bwaidelich** [comment]
I just had another look and we can indeed use `parse_str` instead of Guzzles custom implementation. I replaced the call and added some new test cases covering edge cases and special characters.
Will merge once CI is green
